CP-24290: Remove sm calls & clean up VBDs on termination

gaborigloi opened the pull request:
Not it works in case of service stop, restart, and machine reboot. The only issue is that the client hangs after reboot.

mseri commented on src/main.ml:
```diff
@@ -51,47 +63,52 @@ let require name arg = match arg with
+    (fun () ->
+       Lwt_log.notice_f "Destroying VBD %s" vbd >>= fun () ->
+       Xen_api.VBD.destroy ~rpc ~session_id ~self:vbd >|= fun () ->
+       vbds_to_clean_up := StringSet.remove vbd !vbds_to_clean_up)
```
I don't like accessing references this way, even if it looks safe now. Can you please wrap any change with an `Lwt_mutex.with_lock` (https://ocsigen.org/lwt/2.7.1/api/Lwt_mutex)?

gaborigloi replied:
Sure, thanks for the suggestion!

mseri commented on src/main.ml:
```diff
@@ -51,47 +63,52 @@ let require name arg = match arg with
+let with_vbd ~vDI ~vM ~mode ~rpc ~session_id f =
+  Xen_api.VBD.create ~rpc ~session_id ~vM ~vDI ~userdevice:"autodetect" ~bootable:false ~mode ~_type:`Disk ~unpluggable:true ~empty:false ~other_config:[] ~qos_algorithm_type:"" ~qos_algorithm_params:[]
+  >>= fun vbd ->
+  vbds_to_clean_up := StringSet.add vbd !vbds_to_clean_up;
```
see below

mseri commented on src/main.ml:
```diff
@@ -51,47 +63,52 @@ let require name arg = match arg with
+  | `Error e -> Lwt.fail_with (Printf.sprintf "Unable to read %s: %s" filename (Nbd.Block_error_printer.to_string e))
+  | `Ok b ->
+    let block_uuid = Uuidm.v `V4 |> Uuidm.to_string in
+    Hashtbl.add blocks_to_close block_uuid b;
```
see below

mseri commented on src/main.ml:
```diff
@@ -51,47 +63,52 @@ let require name arg = match arg with
+      (fun () -> f b)
+      (fun () ->
+         Block.disconnect b >|= fun () ->
+         Hashtbl.remove blocks_to_close block_uuid
```
see below

gaborigloi commented:
@mseri Done, now I protect both references with mutexes when I update them.

gaborigloi commented:
One question: should we cause the whole client handler thread to fail when persistent VBD tracking fails? Or should we silently ignore these unexpected I/O errors. I think it might be ok to fail in that case, because we don't expect any errors to occur?

mseri commented:
I'd like @jonludlam and @thomassa to give their opinion, I am not sure how resilient should be the client

mseri commented on src/vbd_store.ml:
```diff
@@ -0,0 +1,37 @@
+    (function
+      | Unix.(Unix_error (EEXIST, "mkdir", dir)) when dir = Consts.xapi_nbd_persistent_dir -> Lwt.return_unit
+      | e ->
+        Lwt_log.error_f "Failed to create directory: %s" (Printexc.to_string e)
```
Now I see what you mean. I think if we have failures here, we should let the client fail and the user/admin fix the permission/space/... issue that caused the failure

mseri commented on src/vbd_store.ml:
```diff
@@ -0,0 +1,37 @@
+let transform_vbd_list f =
+  Lwt_mutex.with_lock m (fun () ->
+      create_dir_if_doesnt_exist () >>= fun () ->
+      (try
```
Why `try` instead of `Lwt.catch`?

gaborigloi replied:
Good spot. `Lwt_stream`s are a bit unique because they are not an `Lwt.t`, so that's why I originally used `try`, but now that I convert them into a list, we do get a `Lwt.t` type back, so we need `Lwt.catch` indeed.

mseri replied:
you mean that depending on the failure we need both try and catch? :O

gaborigloi replied:
`Lwt.catch` is always enough, as it works with normal exceptions too as far as I know :) And we need that in this case, because now we have an Lwt type. Previously I was using one stream piped into a chain of functions, and got a horrible, hard-to-track-down bug where the beginning of the stream was reading what the end of the stream was writing, and it overwrote the original file with duplicate items :face_with_head_bandage:

mseri commented on src/vbd_store.ml:
```diff
@@ -0,0 +1,37 @@
+      create_dir_if_doesnt_exist () >>= fun () ->
+      (try
+         Lwt_io.lines_of_file Consts.vbd_list_file |> Lwt_stream.to_list
+       with _ -> Lwt.return [])
```
This is fine but I'd like to see a log as well

mseri replied:
I don't think we should fail here, I think we should fail on write errors but be resistant to read errors (but with a log when the error is not about non-existent file)

mseri replied:
On the other hand, after discussing it a bit, it may be better to fail to avoid leaking VBDs.

mseri commented on src/vbd_store.ml:
```diff
@@ -0,0 +1,37 @@
+       with _ -> Lwt.return [])
+      >>= fun l ->
+      let l = f l in
+      Lwt_stream.of_list l |> Lwt_io.lines_to_file Consts.vbd_list_file
```
This is not in a `Lwt.catch` either. Although I think we should fail on errors so it seems ok. Let's think if it is worth logging the failure before failing or not. I don't have a strong opinion in regard

mseri commented on src/vbd_store.ml:
```diff
@@ -0,0 +1,37 @@
+  Lwt_unix.file_exists Consts.vbd_list_file >>= fun exists ->
+  if exists then
+    Lwt_mutex.with_lock m (fun () ->
+        Lwt_io.lines_of_file Consts.vbd_list_file |> Lwt_stream.to_list)
```
Here as well, do we want to log the failures?

mseri commented on src/vbd_store.ml:
```diff
@@ -0,0 +1,37 @@
+  transform_vbd_list (List.filter ((<>) vbd_uuid))
+
+let get_all () =
+  Lwt_unix.file_exists Consts.vbd_list_file >>= fun exists ->
```
You are checking this because you know that nothing should delete the file, in the worst case it will be empty. Add a comment explaining this and the fact that this prevents races issues that would require to remove this check and instead use a catch (like you do for the make dir above)

mseri commented on src/vbd_store.mli:
```diff
@@ -0,0 +1,8 @@
+(** Read and write a persistent list of VBD UUIDs.
+    These functions are thread-safe. *)
+
+val add: string -> unit Lwt.t
```
docstrings here please

mseri commented on src/cleanup.ml:
```diff
@@ -0,0 +1,178 @@
+open Lwt.Infix
+
```
Add an interface file here to hide the implementation details

mseri commented:
It would be good to add unit tests. Up to you if you want to add them to this PR or in a separate one, but I expect to see them

gaborigloi commented:
@mseri Thanks for the review, I've addressed your above comments, let me know if more documentation / comments are needed.

mseri commented on src/cleanup.mli:
```diff
@@ -0,0 +1,26 @@
+val ignore_exn_log_error : string -> (unit -> unit Lwt.t) -> unit Lwt.t
+
+module VBD : sig
+  val with_vbd :
```
Please add a docstring

gaborigloi replied:
Done.

mseri commented on src/cleanup.mli:
```diff
@@ -0,0 +1,26 @@
+end
+
+module Block : sig
+  val with_block : string -> (Block.t -> 'a Lwt.t) -> 'a Lwt.t
```
Please add a docstring

gaborigloi replied:
Done.

mseri commented on src/cleanup.mli:
```diff
@@ -0,0 +1,26 @@
+end
+
+module Runtime : sig
+  val register_signal_handler : unit -> unit
```
Please add a docstring. Which signals are handled? What will the signal handler do?

gaborigloi replied:
Done.

mseri commented on src/cleanup.mli:
```diff
@@ -0,0 +1,26 @@
+end
+
+module Persistent : sig
+  val cleanup : unit -> unit Lwt.t
```
Please add a docstring

gaborigloi replied:
Done.

mseri commented on src/vbd_store.ml:
```diff
@@ -1,25 +1,46 @@
       | Unix.(Unix_error (EEXIST, "mkdir", dir)) when dir = Consts.xapi_nbd_persistent_dir -> Lwt.return_unit
       | e ->
-        Lwt_log.error_f "Failed to create directory: %s" (Printexc.to_string e)
+        (* In this we should let the client fail and the user/admin fix the permission/space/... issue that caused the failure *)
```
Reword as
```ocaml
(* In any other case we let the client fail. In this case the user/admin should go and fix the root cause of the issue *)
```

mseri commented on src/vbd_store.ml:
```diff
@@ -1,25 +1,46 @@
       | e ->
-        Lwt_log.error_f "Failed to create directory: %s" (Printexc.to_string e)
+        (* In this we should let the client fail and the user/admin fix the permission/space/... issue that caused the failure *)
+        log_and_reraise_error "Failed to create directory" e
```
Add the path to the error message

mseri commented on src/vbd_store.ml:
```diff
@@ -1,25 +1,46 @@
+        (function
+          | Unix.(Unix_error (ENOENT, "open", file)) when file = vbd_list_file -> Lwt.return []
+          | e ->
+            (* In this we should let the client fail and the user/admin fix the permission/... issue that caused the failure *)
```
Update the comment as above

mseri commented on src/vbd_store.ml:
```diff
@@ -29,9 +50,16 @@ let remove vbd_uuid =
   if exists then
     Lwt_mutex.with_lock m (fun () ->
-        Lwt_io.lines_of_file Consts.vbd_list_file |> Lwt_stream.to_list)
+        Lwt.catch (fun () -> Lwt_io.lines_of_file vbd_list_file |> Lwt_stream.to_list)
```
`<pedantic>`Put the function on a new line as the catches above`</pedantic>`

gaborigloi replied:
You're completely right, I wanted to do that but had to deal with something else so I forgot about it :smile: .

gaborigloi commented:
@mseri Thanks for the review, I think I addressed all the above comments - your comments were completely fair and helpful, actually I just pushed a commit adding all the missing docstrings you requested *before* reading your comments about them :smile: 
Please let me know if any of the comments / docstrings are incorrect / incomplete / unclear.

gaborigloi commented:
@mseri Another downside is that systemd will say the process failed when it's terminated by a signal, because of the exception I throw. But a while ago (https://discuss.ocaml.org/t/lwt-how-to-catch-exceptions-raised-by-signal-handlers/565) I found that the exception is necessary to terminate the program:
```
...
Sep 28 17:43:00 clabuk-02-08 systemd[1]: xapi-nbd.service: main process exited, code=exited, status=1/FAILURE
Sep 28 17:43:00 clabuk-02-08 systemd[1]: Unit xapi-nbd.service entered failed state.
Sep 28 17:43:00 clabuk-02-08 systemd[1]: xapi-nbd.service failed.
```
It's a pity that the solution using `Lwt.wait` + `Lwt.pick` does not work. It worked perfectly for all the small test programs that I've tried. It even worked for parallel background threads as long as I joined them with `Lwt.join`. I tried doing the same thing in `xapi-nbd`, building up a list of background threads and then calling `Lwt.join` on them, but it did not work :(. Maybe the issue is that we have a recursive Lwt thread that does not terminate. Or maybe there is something in the code that I'm missing - not sure.

gaborigloi commented:
I have tested the following scenarios, with different combinations of the various tracking mechanisms:

* Only using `Persistent` tracking:
   * Two clients connecting / disconnecting, the persistent list of VBDs to clean up is the same as the list of VBDs plugged to dom0
  * Two clients connected, stop server with `service xapi-nbd stop`. After termination, the persistent list of VBDs to clean up is the same as the list of VBDs plugged to dom0. After `service xapi-nbd start`, all these VDIs are cleaned up, and both the persistent list of VBDs to clean up and the list of VBDs plugged to dom0 is empty:
  ```
  Sep 28 18:11:33 clabuk-02-08 xapi-nbd[908]: main: Starting xapi-nbd: port = '10809'; certfile = '/etc/xensource/xapi-ssl.pem'; ciphersuites = '!EXPORT:RSA+AES128-SHA256' no_tls = 'false'
  Sep 28 18:11:33 clabuk-02-08 xapi-nbd[908]: main: Checking if there are any VBDs to clean up that leaked during the previous run
  Sep 28 18:11:33 clabuk-02-08 xapi-nbd[908]: main: Cleaning up VBD with UUID c0bfe55f-f201-9db4-80e1-9b2e75dedee1
  Sep 28 18:11:33 clabuk-02-08 xapi-nbd[908]: main: Cleaning up VBD with UUID 347f2673-0b12-d27b-3e98-3546d0fffbef
  Sep 28 18:11:33 clabuk-02-08 xapi-nbd[908]: main: Initialising TLS
  ```
  * Same as above, but with `service xapi-nbd restart`.
  * Same as above, but interrupt with `reboot`.

I used this command to list the VBDs tracked by xapi-nbd, and the VBDs actually plugged to dom0:
```
cat /var/lib/xapi-nbd/VBDs_to_clean_up ; echo ------------------------; xe vbd-list vm-uuid=<UUID of dom0> params=uuid --minimal
```
I used the following command to watch the log output of xapi-nbd:
```
tail -f /var/log/daemon.log | grep xapi-nbd
```

When I did the reboot, I got `ECONNREFUSED`, probably when `xapi-nbd` tried to connect to xapi's Unix domain socket:
```
Sep 28 18:30:33 localhost xapi-nbd[2843]: Caught: Unix.Unix_error(Unix.ECONNREFUSED, "connect", "")
Sep 28 18:30:33 localhost xapi-nbd[2843]: main: Caught unexpected exception: Unix.Unix_error(Unix.ECONNREFUSED, "connect", "")
Sep 28 18:30:33 localhost xapi-nbd[2843]: xapi-nbd: internal error, uncaught exception:
Sep 28 18:30:33 localhost xapi-nbd[2843]: Unix.Unix_error(Unix.ECONNREFUSED, "connect", "")
Sep 28 18:30:33 localhost xapi-nbd[2843]: Raised at file "src/core/lwt.ml", line 805, characters 16-23
Sep 28 18:30:33 localhost xapi-nbd[2843]: Called from file "src/unix/lwt_main.ml", line 34, characters 8-18
Sep 28 18:30:33 localhost xapi-nbd[2843]: Called from file "src/main.ml", line 135, characters 2-181
Sep 28 18:30:33 localhost xapi-nbd[2843]: Called from file "src/cmdliner_term.ml", line 27, characters 19-24
Sep 28 18:30:33 localhost xapi-nbd[2843]: Called from file "src/cmdliner.ml", line 27, characters 27-34
Sep 28 18:30:33 localhost xapi-nbd[2843]: Called from file "src/cmdliner.ml", line 106, characters 32-39
Sep 28 18:30:33 localhost systemd[1]: xapi-nbd.service: main process exited, code=exited, status=1/FAILURE
Sep 28 18:30:33 localhost systemd[1]: Unit xapi-nbd.service entered failed state.
Sep 28 18:30:33 localhost systemd[1]: xapi-nbd.service failed.
Sep 28 18:30:34 localhost systemd[1]: xapi-nbd.service holdoff time over, scheduling restart.
```
When I started it manually later, it worked, and it cleaned up the leaked VDIs properly.
This issue happens consistently at startup: the NBD server will try to make XenAPI calls via the Unix domain socket and will fail.

gaborigloi commented:
By the way, does xapi notify systemd about when its startup sequence finishes? If not, I think that would be a useful thing to add in the future, probably using https://github.com/juergenhoetzel/ocaml-systemd/

gaborigloi commented:
This might be relevant to the solution: https://stackoverflow.com/questions/1372080/wait-for-a-unix-domain-socket-to-be-bound?rq=1
Maybe we can just add a path dependency on the unix domain socket, because systemd path activation uses inotify under the hood, and presumably it listens to `IN_CREATE`. I don't see why it shouldn't work for xapi's unix domain socket, that's also a file after all. If it does not work, we can always add an inotify loop, like this: https://github.com/gaborigloi/xapi-nbd/tree/inotify

Indeed, this page says that Unix domain sockets may not be bound to an existing file, and when they are bound, the file is created: http://osr507doc.sco.com/en/netguide/dusockD.binding_names.html So this means hopefully that we can uses systemd to wait for xapi to become active, because its Unix domain socket will always be freshly created. But this relies on the assumption that this socket is immediately able to receive requests without failures.

gaborigloi commented:
Even if we make the server path-activated by both the unix domain socket of xapi and the `.pem` file, there are still transient errors:
```
Sep 28 21:42:41 localhost xapi-nbd[2605]: Caught: Unix.Unix_error(Unix.ECONNREFUSED, "connect", "")
Sep 28 21:42:41 localhost xapi-nbd[2605]: main: Caught unexpected exception: Unix.Unix_error(Unix.ECONNREFUSED, "connect", "")
Sep 28 21:42:41 localhost xapi-nbd[2605]: xapi-nbd: internal error, uncaught exception:
Sep 28 21:42:41 localhost xapi-nbd[2605]: Unix.Unix_error(Unix.ECONNREFUSED, "connect", "")
Sep 28 21:42:41 localhost xapi-nbd[2605]: Raised at file "src/core/lwt.ml", line 805, characters 16-23
Sep 28 21:42:41 localhost xapi-nbd[2605]: Called from file "src/unix/lwt_main.ml", line 34, characters 8-18
Sep 28 21:42:41 localhost xapi-nbd[2605]: Called from file "src/main.ml", line 135, characters 2-181
Sep 28 21:42:41 localhost xapi-nbd[2605]: Called from file "src/cmdliner_term.ml", line 27, characters 19-24
Sep 28 21:42:41 localhost xapi-nbd[2605]: Called from file "src/cmdliner.ml", line 27, characters 27-34
Sep 28 21:42:41 localhost xapi-nbd[2605]: Called from file "src/cmdliner.ml", line 106, characters 32-39
Sep 28 21:42:41 localhost systemd[1]: xapi-nbd.service: main process exited, code=exited, status=1/FAILURE
Sep 28 21:42:41 localhost systemd[1]: Unit xapi-nbd.service entered failed state.
Sep 28 21:42:41 localhost systemd[1]: xapi-nbd.service failed.
Sep 28 21:42:41 localhost systemd[1]: xapi-nbd.service holdoff time over, scheduling restart.
Sep 28 21:42:41 localhost systemd[1]: Cannot add dependency job for unit lvm2-activation.service, ignoring: Unit lvm2-activation.service is masked.
Sep 28 21:42:41 localhost systemd[1]: Cannot add dependency job for unit lvm2-activation-early.service, ignoring: Unit lvm2-activation-early.service is
masked.
Sep 28 21:42:41 localhost systemd[1]: start request repeated too quickly for xapi-nbd.service
Sep 28 21:42:41 localhost systemd[1]: Failed to start NBD server that exposes XenServer disks.
Sep 28 21:42:41 localhost systemd[1]: Unit xapi-nbd.service entered failed state.
Sep 28 21:42:41 localhost systemd[1]: xapi-nbd.service failed.
Sep 28 21:42:41 localhost ntpd[2475]: 0.0.0.0 c615 05 clock_sync
Sep 28 21:42:45 localhost systemd[1]: Cannot add dependency job for unit lvm2-activation.service, ignoring: Unit lvm2-activation.service is masked.
Sep 28 21:42:45 localhost systemd[1]: Cannot add dependency job for unit lvm2-activation-early.service, ignoring: Unit lvm2-activation-early.service is
masked.
Sep 28 21:42:45 localhost systemd[1]: Started NBD server that exposes XenServer disks.
Sep 28 21:42:45 localhost systemd[1]: Starting NBD server that exposes XenServer disks...
Sep 28 21:42:45 localhost xapi-nbd[2778]: main: Starting xapi-nbd: port = '10809'; certfile = '/etc/xensource/xapi-ssl.pem'; ciphersuites = '!EXPORT:RSA
+AES128-SHA256' no_tls = 'false'
Sep 28 21:42:46 localhost xapi-nbd[2778]: main: Caught unexpected exception: Server_error(INTERNAL_ERROR, [ missing table; host;  ])
Sep 28 21:42:46 localhost xapi-nbd[2778]: xapi-nbd: internal error, uncaught exception:
Sep 28 21:42:46 localhost xapi-nbd[2778]: Server_error(INTERNAL_ERROR, [ missing table; host;  ])
Sep 28 21:42:46 localhost xapi-nbd[2778]: Raised at file "src/core/lwt.ml", line 805, characters 16-23
Sep 28 21:42:46 localhost xapi-nbd[2778]: Called from file "src/unix/lwt_main.ml", line 34, characters 8-18
Sep 28 21:42:46 localhost xapi-nbd[2778]: Called from file "src/main.ml", line 135, characters 2-181
Sep 28 21:42:46 localhost xapi-nbd[2778]: Called from file "src/cmdliner_term.ml", line 27, characters 19-24
Sep 28 21:42:46 localhost xapi-nbd[2778]: Called from file "src/cmdliner.ml", line 27, characters 27-34
Sep 28 21:42:46 localhost xapi-nbd[2778]: Called from file "src/cmdliner.ml", line 106, characters 32-39
Sep 28 21:42:46 localhost systemd[1]: xapi-nbd.service: main process exited, code=exited, status=1/FAILURE
Sep 28 21:42:46 localhost systemd[1]: Unit xapi-nbd.service entered failed state.
Sep 28 21:42:46 localhost systemd[1]: xapi-nbd.service failed.
Sep 28 21:42:46 localhost systemd[1]: xapi-nbd.service holdoff time over, scheduling restart.
Sep 28 21:42:46 localhost systemd[1]: Started NBD server that exposes XenServer disks.
Sep 28 21:42:46 localhost systemd[1]: Starting NBD server that exposes XenServer disks...
Sep 28 21:42:46 localhost xapi-nbd[2805]: main: Starting xapi-nbd: port = '10809'; certfile = '/etc/xensource/xapi-ssl.pem'; ciphersuites = '!EXPORT:RSA+AES128-SHA256' no_tls = 'false'
Sep 28 21:42:46 localhost xapi-nbd[2805]: main: Checking if there are any VBDs to clean up that leaked during the previous run
Sep 28 21:42:46 localhost xapi-nbd[2805]: main: Initialising TLS
Sep 28 21:42:46 localhost xapi-nbd[2805]: main: Setting up server socket
Sep 28 21:42:46 localhost xapi-nbd[2805]: main: Listening for incoming connections
```
So it seems that xapi is not yet completely ready when its Unix domain socket is created (`Sep 28 21:42:46 localhost xapi-nbd[2778]: Server_error(INTERNAL_ERROR, [ missing table; host;  ])`). So I'll just add a loop to the startup of xapi-nbd that will try logging in until it succeeds, up to a timeout, to ensure that xapi is up when we need it.

mseri commented:
I'd like to see this discussion in a `issue` and the fix in a different PR, this one is already big enough, and the socket mechanism predates it anyway

gaborigloi commented on src/cleanup.ml:
```diff
@@ -0,0 +1,180 @@
+
+module Persistent = struct
+  let cleanup () =
+    local_login () >>= fun (rpc, session_id) ->
```
One thing that I'm not sure about is whether I should log out of these local sessions. It's not a disaster if I don't do that, xapi would just invalidate the oldest sessions of the server, because we supply the `originator` name when we log in.

mseri replied:
I think it better to log out

gaborigloi replied:
Will do in the the PR.

gaborigloi commented:
@mseri raised https://github.com/xapi-project/xapi-nbd/pull/20 to fix the startup and log out of the local session

gaborigloi commented:
Added a unit test for the `Vbd_store` in https://github.com/xapi-project/xapi-nbd/pull/21
